src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Hierarchy.cs

```diff
@@ -184,405 +192,206 @@ public static partial class SymbolFinder
         {
-            // A symbol can only have implementations if it's an interface or a
-            // method/property/event from an interface.
-            if (symbol is INamedTypeSymbol namedTypeSymbol)
-            {
-                var implementingTypes = await DependentTypeFinder.FindTransitivelyImplementingStructuresAndClassesAsync(
-                    namedTypeSymbol, solution, projects, cancellationToken).ConfigureAwait(false);
-                return ImmutableArray<ISymbol>.CastUp(implementingTypes).WhereAsArray(IsAccessible);
-            }
-            else if (symbol.IsImplementableMember())
-            {
-                var containingType = symbol.ContainingType.OriginalDefinition;
-                var allTypes = await DependentTypeFinder.FindTransitivelyImplementingStructuresClassesAndInterfacesAsync(
-                    containingType, solution, projects, cancellationToken).ConfigureAwait(false);
-
-                ImmutableArray<ISymbol>.Builder results = null;
-                foreach (var t in allTypes)
-                {
-                    var implementations = await t.FindImplementationsForInterfaceMemberAsync(symbol, solution, cancellationToken).ConfigureAwait(false);
-                    foreach (var implementation in implementations)
-                    {
-                        var sourceDef = await FindSourceDefinitionAsync(implementation, solution, cancellationToken).ConfigureAwait(false);
-                        var bestDef = sourceDef ?? implementation;
-                        if (IsAccessible(bestDef))
-                        {
-                            results ??= ImmutableArray.CreateBuilder<ISymbol>();
-                            results.Add(bestDef.OriginalDefinition);
-                        }
-                    }
-                }
-
-                if (results != null)
-                {
-                    return results.Distinct(SymbolEquivalenceComparer.Instance).ToImmutableArray();
-                }
-            }
+            if (type == null)
+                throw new ArgumentNullException(nameof(type));
 
-            return ImmutableArray<ISymbol>.Empty;
-        }
+            if (solution == null)
+                throw new ArgumentNullException(nameof(solution));
 
-        private static bool IsAccessible(ISymbol symbol)
-        {
-            if (symbol.Locations.Any(l => l.IsInMetadata))
-            {
-                var accessibility = symbol.DeclaredAccessibility;
-                return accessibility == Accessibility.Public ||
-                    accessibility == Accessibility.Protected ||
-                    accessibility == Accessibility.ProtectedOrInternal;
-            }
+            if (solution.GetOriginatingProjectId(type) == null)
+                throw new ArgumentException(WorkspacesResources.Symbols_project_could_not_be_found_in_the_provided_solution, nameof(type));
```

Comment:
note: all public entrypoints got beefed up in terms of actually checking their args.